src/components/SourcesTree.js

```diff
@@ -44,6 +57,33 @@ let SourcesTree = React.createClass({
   },
 
   componentWillReceiveProps(nextProps) {
+    if (isEnabled("showSource") &&
+    nextProps.shownSource != this.props.shownSource) {
+      const tempList = [];
+      const sourceURL = nextProps.shownSource;
+      const sourceTreeList = this.state.sourceTree;
+      const itemsStrings = returnItemsStrings(sourceURL);
+      const processIndex = 0;
+
+      // Determine which item(s) should be added to itemsList
+      while (processIndex < itemsStrings.length) {
+        const i = 0;
+        const found = false;
+        while (!found) {
+          if (itemsStrings[processIndex] == sourceTreeList.contents[i].path) {
+            tempList.push(sourceTreeList.contents[i]);
+            sourceTreeList = sourceTreeList.contents[i];
+            found = true;
+            processIndex++;
+          }
+          i++;
+        }
+      }
```

Comment:
I don't think that using while loops is a good idea - it's difficult to follow the indices and it might as well run forever.

I've never run this code but the `const` reassignments look very suspicious...